### PR TITLE
Use Winston for logging

### PR DESCRIPTION
**Pull request by giordano**
Address #26 and #33.  Alternative to #27: we use Winston directly so we have
full control of the logging format.

### CODE REVIEW

File: src/logger.ts [hunk 1]
```diff
@@ -0,0 +1,93 @@
+import * as os from 'os';
+import * as path from 'path';
+import * as vscode from 'vscode';
+import * as winston from 'winston';
+import * as TransportStream from 'winston-transport';
+
+class OutputChannelTransport extends TransportStream {
```
Comment (ageorgou) [inline src/logger.ts:7]: If we took this without modification from somewhere, can we link to the source (for credit and also future maintenability)?

File: src/logger.ts [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+let fileLogger;
+let consoleLogger;
+let isInitialised: Boolean = false;
+
+export function initLogging(outputChannel: vscode.OutputChannel, filename?: string) {
+    const format = winston.format.combine(
+        winston.format.label({ label: 'nisaba' }),
+        winston.format.timestamp(),
+        fileLoggerFormat
+    );
```
Comment (ageorgou) [inline src/logger.ts:41]: So this is like a partial application of the format? Neat!

File: src/logger.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    consoleLogger = winston.createLogger({
+        level: 'info',
+        format: format,
+        defaultMeta: { service: 'user-service' },
+        transports: [
+            new OutputChannelTransport(outputChannel),
+        ],
+    });
+
+    isInitialised = true;
```
Comment (ageorgou) [inline src/logger.ts:69]: Should the whole body be skipped if `isInitialised` is already true? Do we want to warn (for our benefit)? In case we do more intricate logging in the future...

File: src/test/suite/grammar.test.ts [hunk 4]
```diff
@@ -37,8 +36,6 @@ const registry = new vsctm.Registry({
 });
 
 suite('Grammar Test Suite', () => {
-    vscode.window.showInformationMessage('Start all tests.');
```
Comment (ageorgou) [inline src/test/suite/grammar.test.ts:40]: This is unrelated, right?

File: src/test/suite/logger.test.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
+// import * as assert from 'assert';
+import * as vscode from 'vscode';
+import * as fs from 'fs';
+import * as os from 'os';
+import * as tmp from 'tmp';
+import * as assert from 'assert';
+import * as nisabaLogger from '../../logger';
+
+const nisabaOutputChannel = vscode.window.createOutputChannel("Nisaba");
```
Comment (ageorgou) [inline src/test/suite/logger.test.ts:9]: Very minor: is it worth having this as a separate function (e.g. `createOutputChannel`) so we can use the actual variable from the `extension.ts` without calling all of `activate`?

File: src/test/suite/logger.test.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+// Log messages to the temporary file.
+const messages = [
+    ['debug', 'This is a debug message'],
+    ['info', 'This is an info'],
+    ['warn', 'This is a warning'],
+    ['error', 'This is an error']
+]
+for (let i = 0; i < messages.length; i++) {
+    nisabaLogger.log(messages[i][0], messages[i][1]);
+}
```
Comment (ageorgou) [inline src/test/suite/logger.test.ts:24]: Not crucial, but we can also use a more functional interface with arrays:
```suggestion
messages.forEach(message => nisabaLogger.log(message[0], message[1]))
```

or even
```suggestion
messages.forEach(([level, message]) => nisabaLogger.log(level, message))
```

**Review by ageorgou — COMMENTED** [review]
Looks good! We're not testing the console output because we can't read from the output channel, right? I wonder if there an easy way to mock that?

File: src/test/suite/grammar.test.ts (hunk 4 above)
Comment (giordano) [reply src/test/suite/grammar.test.ts:40]: Kinda unrelated: I spotted it when creating the logger file test.  This message is pretty much useless, probably taken from an example test.  I thought it wouldn't be _too_ bad to remove these lines here, this won't even hurt git-blaming :upside_down_face:

**Comment by ageorgou** [comment]
Actually should have approved this, many of the comments are simply suggestsions and there is nothing that worries me.

**Review by ageorgou — APPROVED** [review]

File: src/logger.ts (hunk 1 above)
Comment (giordano) [reply src/logger.ts:7]: I think this is based on Winston documentation about how to define custom transports: https://github.com/winstonjs/winston/blob/2625f60c5c85b8c4926c65e98a591f8b42e0db9a/README.md#adding-custom-transports.  The extra bit is that here the class has an attribute

Comment (ageorgou) [reply src/logger.ts:7]: Okay, I thought it had come from another example for some reason!

File: src/logger.ts (hunk 3 above)
Comment (giordano) [reply src/logger.ts:75]: Done!

File: src/test/suite/logger.test.ts (hunk 5 above)
Comment (giordano) [reply src/test/suite/logger.test.ts:9]: This variable lives in this module, so there is no need to go through `activate`.  However we can't really do much with an [`OutputChannel`](https://code.visualstudio.com/api/references/vscode-api#OutputChannel), in particular we can't access its content, which would be useful for testing.
